### PR TITLE
OpenAI conversation:CRUDs

### LINKED ISSUE #298: OpenAI Conversation
**Is your feature request related to a problem? Please describe.**
We’re seeing an issue with the Response API. The conversation_id seems to change with every message.

According to the docs here:
https://platform.openai.com/docs/guides/conversation-state?api-mode=responses#openai-apis-for-conversation-state

If we extend the code to send a third message, we’re supposed to pass the response_id from the second_response. But this response_id is different from the response_id from first_response — it has the same prefix and suffix, but the middle part changes.

This seems quite different from the Threads API, where one can keep the same thread_id for the whole conversation, even though each message has its own message_id.

**Describe the solution you'd like**
Thus to mitigate this issue we have to store conversation history at our end and to do that first need to create a new table and have basic CRUD operations

### PR DESCRIPTION
**Pull request by AkhileshNegi**
## Summary

Target issue is #298
Explain the **motivation** for making this change. What existing problem does the pull request solve?

## Checklist

Before submitting a pull request, please ensure that you mark these task.

- [x] Ran `fastapi run --reload app/main.py` or `docker compose up` in the repository root and test.
- [x] If you've fixed a bug or added code that is tested and has test cases.

## Notes

Please add here if any other information is required for the reviewer.

## Summary by CodeRabbit

* **New Features**
  * Introduced API endpoints to retrieve, list (with pagination), and soft-delete OpenAI conversation records within a project.
  * Added support for managing OpenAI conversation data, including threading, project/organization scoping, and timestamp tracking.
  * Implemented validation for OpenAI response ID formats.
  * Added database support with a new table for OpenAI conversations, including indexing and foreign key constraints.

* **Bug Fixes**
  * Not applicable.

* **Tests**
  * Added comprehensive tests for API endpoints and CRUD operations related to OpenAI conversations, including edge cases and pagination.

* **Chores**
  * Added utility functions for generating OpenAI-style IDs and retrieving organizations for testing purposes.

### CODE REVIEW

File: backend/app/models/openai_conversation.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    if v is None:
+        return v
+    if not re.match(r"^resp_[a-zA-Z0-9]{10,}$", v):
+        raise ValueError(
+            "response_id fields must follow pattern: resp_ followed by at least 10 alphanumeric characters"
+        )
+    return v
+
+
+def validate_assistant_id_pattern(v: str) -> str:
```
Comment (avirajsingh7) [inline backend/app/models/openai_conversation.py]: we have currently added a endpoint, crud for assistant,
There we are not interacting with openAI, so when new assistant is created we are using uuid,
That's why we need to remove this validation

File: backend/app/models/openai_conversation.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        raise ValueError(
+            "assistant_id must follow pattern: asst_ followed by at least 10 alphanumeric characters"
+        )
+    return v
+
+
+class OpenAIConversationBase(SQLModel):
+    # usually follow the pattern of resp_688704e41190819db512c30568xxxxxxx
+    response_id: str = Field(index=True, min_length=10)
+    ancestor_response_id: Optional[str] = Field(
```
Comment (avirajsingh7) [inline backend/app/models/openai_conversation.py]: This should not be nullable, As we discussed, right?

File: backend/app/api/routes/openai_conversation.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    OpenAIConversation,
+)
+from app.utils import APIResponse
+
+router = APIRouter(prefix="/openai-conversation", tags=["OpenAI Conversations"])
+
+
+@router.get(
+    "/{conversation_id}",
+    response_model=APIResponse[OpenAIConversation],
```
Comment (avirajsingh7) [inline backend/app/api/routes/openai_conversation.py]: This is something I missed in my code also
Here return response should be OpenAIConversationPublic 
things like is_deleted and deleted_at should not be visible to user
and let say in future if add some more field to table and do not want to send back then to user,
Than this will create a problem

File: backend/app/crud/openai_conversation.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+) -> OpenAIConversation | None:
+    """
+    Return a conversation by its ancestor response ID and project.
+    """
+    statement = select(OpenAIConversation).where(
+        OpenAIConversation.ancestor_response_id == ancestor_response_id,
+        OpenAIConversation.project_id == project_id,
+        OpenAIConversation.is_deleted == False,
+    )
+    result = session.exec(statement).first()
```
Comment (avirajsingh7) [inline backend/app/crud/openai_conversation.py]: A  ancestor can have multiple conversation but here it is returning only one,
We should return all the conversation of that particular ancestor

File: backend/app/api/routes/openai_conversation.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    skip: int = Query(0, ge=0, description="How many items to skip"),
+    limit: int = Query(100, ge=1, le=100, description="Maximum items to return"),
+):
+    """
+    List all conversations in the current project.
+    """
+    conversations = get_conversations_by_project(
+        session=session, project_id=current_user.project_id, skip=skip, limit=limit
+    )
+    return APIResponse.success_response(conversations)
```
Comment (avirajsingh7) [inline backend/app/api/routes/openai_conversation.py]: can we start including pagination metadata in list endpoint,
values such as skip, limit and total
```
        "metadata": {
            "skip": skip,
            "limit": limit,
            "total": total
        }
```

File: backend/app/tests/api/routes/test_openai_conversation.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+
+def test_get_conversation_success(
+    client: TestClient,
+    db: Session,
+    user_api_key_header: dict,
+):
+    """Test successful conversation retrieval."""
+    # Get the project ID from the user's API key
+    from app.tests.utils.utils import get_user_from_api_key
```
Comment (avirajsingh7) [inline backend/app/tests/api/routes/test_openai_conversation.py]: import should not be here

File: backend/app/tests/utils/conversation.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            select(OpenAIConversation)
+            .where(OpenAIConversation.is_deleted == False)
+            .limit(1)
+        )
+
+    conversation = session.exec(statement).first()
+
+    if not conversation:
+        # Create a new conversation if none exists
+        from app.tests.utils.utils import get_project, get_organization
```
Comment (avirajsingh7) [inline backend/app/tests/utils/conversation.py]: import should not be here

File: backend/app/tests/utils/conversation.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                select(Project).where(Project.id == project_id)
+            ).first()
+            if not project:
+                raise ValueError(f"Project with ID {project_id} not found")
+        else:
+            project = get_project(session)
+
+        organization = get_organization(session)
+
+        conversation_data = OpenAIConversationCreate(
```
Comment (avirajsingh7) [inline backend/app/tests/utils/conversation.py]: create is different functionality and get is different functionality,
We should not mix them by using them in same method

**Review by avirajsingh7 — CHANGES_REQUESTED** [review]

File: backend/app/models/openai_conversation.py (hunk 2 above)
Comment (AkhileshNegi) [reply backend/app/models/openai_conversation.py]: updated

File: backend/app/api/routes/openai_conversation.py (hunk 3 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/openai_conversation.py]: updated

File: backend/app/tests/api/routes/test_openai_conversation.py (hunk 6 above)
Comment (AkhileshNegi) [reply backend/app/tests/api/routes/test_openai_conversation.py]: updated

File: backend/app/api/routes/openai_conversation.py (hunk 5 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/openai_conversation.py]: added

File: backend/app/crud/openai_conversation.py (hunk 4 above)
Comment (AkhileshNegi) [reply backend/app/crud/openai_conversation.py]: we'll return the latest response for an ancestor id as that's what we'll need in conversation memory

File: backend/app/models/openai_conversation.py (hunk 1 above)
Comment (AkhileshNegi) [reply backend/app/models/openai_conversation.py]: removed

**Review by avirajsingh7 — CHANGES_REQUESTED** [review]
When running this command `alembic revision --autogenerate -m "Test"`, I was able to see some changes in new version of alembic file.
This command should not generate anything, if it generate this shows there is some difference between current sql model and db state

Please check this
